src/ssl.c

```diff
@@ -25066,35 +25104,80 @@ WOLFSSL_API int wolfSSL_CTX_set1_curves_list(WOLFSSL_CTX* ctx, char* names)
 #endif
 
 #ifdef OPENSSL_EXTRA
+#ifndef NO_WOLFSSL_STUB
 int wolfSSL_CTX_set_msg_callback(WOLFSSL_CTX *ctx, SSL_Msg_Cb cb)
 {
     WOLFSSL_STUB("SSL_CTX_set_msg_callback");
     (void)ctx;
     (void)cb;
     return SSL_FAILURE;
 }
+#endif
+
 int wolfSSL_set_msg_callback(WOLFSSL *ssl, SSL_Msg_Cb cb)
 {
-    WOLFSSL_STUB("SSL_set_msg_callback");
-    (void)ssl;
-    (void)cb;
-    return SSL_FAILURE;
+  WOLFSSL_ENTER("wolfSSL_set_msg_callback");
+
+  if (ssl == NULL) {
+      return SSL_FAILURE;
+  }
+
+  if (cb != NULL) {
+      ssl->toInfoOn = 1;
+  }
+
+  ssl->protoMsgCb = cb;
+  return SSL_SUCCESS;
 }
-int wolfSSL_CTX_set_msg_callback_arg(WOLFSSL_CTX *ctx, void* arg)
+
+#ifndef NO_WOLFSSL_STUB
+void wolfSSL_CTX_set_msg_callback_arg(WOLFSSL_CTX *ctx, void* arg)
 {
     WOLFSSL_STUB("SSL_CTX_set_msg_callback_arg");
     (void)ctx;
     (void)arg;
-    return SSL_FAILURE;
+    return;
 }
-int wolfSSL_set_msg_callback_arg(WOLFSSL *ssl, void* arg)
+#endif
+
+void wolfSSL_set_msg_callback_arg(WOLFSSL *ssl, void* arg)
 {
-    WOLFSSL_STUB("SSL_set_msg_callback_arg");
-    (void)ssl;
-    (void)arg;
-    return SSL_FAILURE;
+    WOLFSSL_ENTER("wolfSSL_set_msg_callback_arg");
+    ssl->protoMsgCtx = arg;
+    return;
 }
-#endif
 
+void *wolfSSL_OPENSSL_memdup(const void *data, size_t siz, const char* file, int line)
+{
+    (void)file;
+    (void)line;
+    void *ret;
+
+    if (data == NULL || siz >= INT_MAX)
+        return NULL;
+
+    ret = OPENSSL_malloc(siz);
+    if (ret == NULL) {
+        return NULL;
+    }
+    return XMEMCPY(ret, data, siz);
+}
+
+int wolfSSL_CTX_set_alpn_protos(WOLFSSL_CTX *ctx, const unsigned char *p,
+                            unsigned int p_len)
+{
+    if((void *)ctx->alpn_cli_protos != NULL)
```

Comment:
Shouldn't this function should check "ctx" and "p" arguments for NULL pointers and return BAD_FUNC_ARG, to prevent against crashes in those cases?